src/ExtraValuesAssigner.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace Wikibase\DataModel;
+
+/**
+ * Assigns extra values to the serialization of an object
+ * based on the given extra values providers.
+ *
+ * @since 2.1
+ *
+ * @licence GNU GPL v2+
+ * @author Bene* < benestar.wikimedia@gmail.com >
+ */
+class ExtraValuesAssigner {
+
+	/**
+	 * @var ExtraValuesProvider[]
+	 */
+	private $extraValuesProviders;
+
+	/**
+	 * @param ExtraValuesProvider[] $extraValuesProviders
+	 */
+	public function __construct( array $extraValuesProviders = array() ) {
+		$this->extraValuesProviders = $extraValuesProviders;
+	}
+
+	/**
+	 * @param array $serialization
+	 * @param mixed $object
+	 * @return array
+	 */
+	public function addExtraValues( $serialization, $object ) {
+		foreach ( $this->extraValuesProviders as $extraValuesProvider ) {
+			if ( $extraValuesProvider->isProviderFor( $object ) ) {
+				$serialization = array_merge(
+					$extraValuesProvider->getExtraValues( $object ),
+					$serialization
```

Comment:
I want a clear cut. The injection code can even go into it's own component, independent from anything but this one. My problem with the approach in this pull request is that:

1. It encodes knowledge about the places **where** stuff can be injected into this component.
2. But **what** can be injected and **how** it must be serialized and deserialized is outside of this component.

I do not think the first issue is the biggest we need to solve. Hard-code some paths. So what? They can not change anyway without becoming a big fat breaking change. This can easily be solved by making the injection component depend on this one.

The problem with the approach in this patch is that you can have breaking changes in the serialization in two different places. If you change something in one of the "extra value" providers it immediately creates a breaking change, but there is no composer dependency in that direction.